## Introduction

## Project - Finance

Predict whether a credit card user will default on monthly credit card payment based on annual income and monthly credit card balance

## Project Steps Followed

1. Define Project Goals/Objective
2. Data Retrieval
3. Data Cleansing
4. Exploratory Data Analysis
5. Data Modeling
6. Result Analysis

## About the Project

We have information about credit card balance and annual income for 10,000  individuals. Based on the data, we need to predict whether the individual will default on a monthly credit card balance

## Define Research Goals

Predict whether a credit card user will default on monthly credit card payment based on annual income and monthly credit card balance

## Data Set

We have data about credit card balance and annual income for 10,000 individuals

## Load Data and Analysis

In [ ]:
import pandas as pd

In [182]:
import matplotlib.gridspec as gridspec

In [183]:
import numpy as np

In [184]:
import matplotlib as mpl

In [185]:
import matplotlib.pyplot as plt

In [186]:
import seaborn as sns

In [195]:
plt.style.use('seaborn-white')

C:\Users\mukes\AppData\Local\Temp\ipykernel_13620\1193657973.py:1: MatplotlibDeprecationWarning: The seaborn styles shipped by Matplotlib are deprecated since 3.6, as they no longer correspond to the styles shipped by seaborn. However, they will remain available as 'seaborn-v0_8-<style>'. Alternatively, directly use the seaborn API instead.
  plt.style.use('seaborn-white')


In [196]:
ccdef = pd.read_excel('Default.xlsx')

C:\Users\mukes\anaconda3\Lib\site-packages\openpyxl\styles\stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [197]:
ccdef.head()

Unnamed: 0 default student      balance        income
0           1      No      No   729.526495  44361.625074
1           2      No     Yes   817.180407  12106.134700
2           3      No      No  1073.549164  31767.138947
3           4      No      No   529.250605  35704.493935
4           5      No      No   785.655883  38463.495879

In [198]:
ccdef.shape

(10000, 5)

In [199]:
ccdef.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  10000 non-null  int64  
 1   default     10000 non-null  object 
 2   student     10000 non-null  object 
 3   balance     10000 non-null  float64
 4   income      10000 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 390.8+ KB


Our observations are as follows

1. NaN values do not present in the data set. Because of the Non-Null Count and number of rows in the dataset match.
2. There are 3 Input Variables and 1 Output Variable (default)
3. The data type of balance and income variables is float64. The data type of out variable (default) and student is object.
4. Shows two input variables continuous (quantitative) data types.
5. Output variable as well as 1 input variable (student) are categorical (qualitative) data types
6. None of the columns contain the Null Values

In [200]:
ccdef.isnull().sum()

Unnamed: 0    0
default       0
student       0
balance       0
income        0
dtype: int64

## Exploratory Data Analysis

### Statistical Analysis

In [201]:
ccdef.describe()

Unnamed: 0       balance        income
count  10000.00000  10000.000000  10000.000000
mean    5000.50000    835.374886  33516.981876
std     2886.89568    483.714985  13336.639563
min        1.00000      0.000000    771.967729
25%     2500.75000    481.731105  21340.462903
50%     5000.50000    823.636973  34552.644802
75%     7500.25000   1166.308386  43807.729272
max    10000.00000   2654.322576  73554.233495

We can see that the min value of balance is zero. We need to confirm how many zero values existing in the dataset.

For all other columns, the data cleaning is not required. However for categorical variables, the encoding is required

In [202]:
(ccdef.balance == 0).sum(axis=0)

499

499 rows of the balance variable contain the zero value, which is possible. Hence we conclude the data cleaning steps are not required for the balance variabl

### Categorical Variable Analysis

In [203]:
ccdef.student.value_counts()

student
No     7056
Yes    2944
Name: count, dtype: int64

This confirms that the predictor student has only 2 possible values. Yes and No. The distribution of students vs non-students is given above.

### Response Variable Analysis

In [204]:
ccdef.default.value_counts()

default
No     9667
Yes     333
Name: count, dtype: int64

This confirms that the response variable default has only 2 possible values. Yes and No. Data is highly skewed. Only  of the individuals in training data defaulted.

### Encode Categorical Variables

Most machine learning models accept the numerical data only. It is necessary to pre-process the categorical variables. We need to convert the categorical variables into numbers. For any machine learning project, converting categorical data is an unavoidable activity.

In [205]:
ccdef['default2'] = ccdef.default.factorize()[0]

In [206]:
ccdef['student2'] = ccdef.default.factorize()[0]

In [207]:
ccdef.head(3)

Unnamed: 0 default student      balance        income  default2  student2
0           1      No      No   729.526495  44361.625074         0         0
1           2      No     Yes   817.180407  12106.134700         0         0
2           3      No      No  1073.549164  31767.138947         0         0

### Graphical Representation


Relationship between balance and income and the relationship between default and balance, and default and income, has been plotted.

We create a new data frame, ccdef_df, that includes 15% data for non defaulters and whole data for defaulters

In [ ]:
ccdef_df = ccdef[ccdef.default2 == 0].sample(frac=0.15)

In [228]:
ccdef_dfyes = ccdef[ccdef.default2 == 1]

In [229]:
result_df = pd.concat([ccdef_dfno, ccdef_dfyes], axis=0)

In [236]:
%matplotlib

Using matplotlib backend: QtAgg


In [237]:
fig = plt.figure(figsize=(12,5))

In [238]:
gs = mpl.gridspec.GridSpec(1, 4)

In [239]:
ax1 = plt.subplot(gs[0,:2])

In [240]:
ax2 = plt.subplot(gs[0,2:3])

In [241]:
ax3 = plt.subplot(gs[0,3:4])

In [250]:
ax1.scatter(ccdef_df[ccdef_df.default == 'Yes'].balance, ccdef_df[ccdef_df.default == 'Yes'].income, s=40, c='orange', marker='+', linewidths=1) 

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [251]:
ax1.scatter(ccdef_df[ccdef_df.default == 'No'].balance, ccdef_df[ccdef_df.default == 'No'].income, s=40, marker='o', linewidths='1', edgecolors='lightblue', facecolors='white', alpha=.6) 

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [252]:
ax1.set_ylim(ymin=0)

(0.0, 73387.58174073232)

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [253]:
ax1.set_ylabel('Income')

Text(125.72916666666667, 0.5, 'Income')

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [254]:
ax1.set_xlim(xmin=-100) 

(-100.0, 2224.7406029300814)

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [255]:
ax1.set_xlabel('Balance')

Text(0.5, 40.72916666666666, 'Balance')

Traceback (most recent call last):
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_qt.py", line 468, in _draw_idle
    self.draw()
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\backends\backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\figure.py", line 3175, in draw
    mimage._draw_list_compositing_images(
  File "C:\Users\mukes\AppData\Roaming\Python\Python311\site-packages\matplotlib\image.py", li

In [256]:
c_palette = {'No':'lightblue', 'Yes':'orange'}

In [257]:
sns.boxplot('default', 'balance', data=ccdef, orient='v', ax=ax2, palette=c_palette) 

TypeError: boxplot() got multiple values for argument 'data'

In [ ]:
sns.boxplot('default', 'income', data=ccdef, orient='v', ax=ax3, palette=c_palette) 

In [258]:
gs.tight_layout(plt.gcf())